runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -248,31 +250,37 @@ private StateRequestHandler getStateRequestHandler(ExecutableStage executableSta
     return StateRequestHandlers.delegateBasedUponType(handlerMap);
   }
 
-  private static class BagUserStateFactory<K extends ByteString, V, W extends BoundedWindow>
+  static class BagUserStateFactory<K extends ByteString, V, W extends BoundedWindow>
       implements StateRequestHandlers.BagUserStateHandlerFactory<K, V, W> {
 
     private final StateInternals stateInternals;
     private final KeyedStateBackend<ByteBuffer> keyedStateBackend;
     private final Lock stateBackendLock;
+    /** Holds the current valid cache token for this operator. */
+    private final ByteString cacheToken;
 
-    private BagUserStateFactory(
+    BagUserStateFactory(
         StateInternals stateInternals,
         KeyedStateBackend<ByteBuffer> keyedStateBackend,
         Lock stateBackendLock) {
 
       this.stateInternals = stateInternals;
       this.keyedStateBackend = keyedStateBackend;
       this.stateBackendLock = stateBackendLock;
+      this.cacheToken = ByteString.copyFrom(UUID.randomUUID().toString(), Charsets.UTF_8);
```

Comment:
nit: Please take in an `IdGenerator` as a parameter since it will allow one to explicitly test knowing what the cache token value(s) will be. Also consider using `IdGenerators#incrementingLongs`

Reply:
I had the assumption that the cache token would be globally unique. That avoids the cleanup scenario when dealing with a long-running SDK Harness. You are right though that testing/debugging is much easier with increasing integers. Have made the cache token generation configurable via `IdGenerator`.